### PR TITLE
Add mask target generator operator for Mask-RCNN

### PR DESCRIPTION
**Pull request by Kh4L**
## Description ##
Add `mrcnn_target` GPU operator.

It computes Mask-RCNN mask targets. It takes the same arguments and produces the same outputs as [this Gluon block](https://github.com/dmlc/gluon-cv/blob/master/gluoncv/model_zoo/mask_rcnn/rcnn_target.py). 

### Micro-benchmark:
**mrcnn_target operator** 374ns vs **Reference** 5.43μs

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Operator, tests, (and when applicable, API doc)

## Comments ##
- [x] Add type support
- [ ] Need to test Symbol API

### CODE REVIEW

**Comment by anirudhacharya** [comment]
> ### Micro-benchmark:
> **mrcnn_target operator** 374ns vs **Reference** 5.43μs

What is the reference in this benchmark?

File: src/operator/contrib/mrcnn_target.cu [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                                  const int total_out_el,
+                                  int batch_size,
+                                  int num_classes,
+                                  int num_rois,
+                                  int num_gtmasks,
+                                  int gt_height,
+                                  int gt_width,
+                                  int m) {
+  // computing sampled_masks
+  const int sampling_ratio = 2;
```
Comment (wkcn) [inline src/operator/contrib/mrcnn_target.cu]: Thank you for the great contribution!
Could you please add sampling_ratio into MRCNNTargetParam?

File: src/operator/contrib/mrcnn_target.cu [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+                                  DType* sampled_masks,
+                                  DType* mask_cls,
+                                  const int total_out_el,
+                                  int batch_size,
+                                  int num_classes,
+                                  int num_rois,
+                                  int num_gtmasks,
+                                  int gt_height,
+                                  int gt_width,
+                                  int m) {
```
Comment (wkcn) [inline src/operator/contrib/mrcnn_target.cu]: A meaningful variable name will be better : )

**Comment by Jerryzcn** [comment]
> > ### Micro-benchmark:
> > **mrcnn_target operator** 374ns vs **Reference** 5.43μs
> 
> What is the reference in this benchmark?

Probably the current python version

File: src/operator/contrib/mrcnn_target.cu [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  for (int mask_idx = blockIdx.x; mask_idx < num_masks; mask_idx += gridDim.x) {
+    int cls_idx = mask_idx % num_classes;
+    int roi_idx = (mask_idx / num_classes) % num_rois;
+    int batch_idx = (mask_idx / num_classes / num_rois);
+
+    DType* mask_cls_out = mask_cls + mask_idx * mask_vol;
+
+    DType cls_target = cls_targets[batch_idx * num_rois + roi_idx];
+    for (int mask_pixel = threadIdx.x; mask_pixel < mask_vol; mask_pixel += blockDim.x) {
+      mask_cls_out[mask_pixel] = (cls_target == cls_idx);
```
Comment (wkcn) [inline src/operator/contrib/mrcnn_target.cu]: I am not sure whether “ (cls_target == cls_idx)” will be optimized by the compiler.

**Comment by Kh4L** [comment]
> > **mrcnn_target operator** 374ns vs **Reference** 5.43μs

@anirudhacharya 

@Jerryzcn is right, it is [this Gluon code](https://github.com/dmlc/gluon-cv/blob/master/gluoncv/model_zoo/mask_rcnn/rcnn_target.py), hybridized.

File: src/operator/contrib/mrcnn_target.cu (hunk 3 above)
Comment (Kh4L) [reply src/operator/contrib/mrcnn_target.cu]: Are you talking about the implicit casting?
I think it is ok, but I can also change to

```suggestion
      mask_cls_out[mask_pixel] = static_cast<DType>(static_cast<int>(cls_target) == cls_idx);
```

Comment (wkcn) [reply src/operator/contrib/mrcnn_target.cu]: No. Sorry that I did not describe it clearly.
` (cls_target == cls_idx)` is a constant value in the loop. I do not know whether the compiler will consider it as a constant value like that:
```c++
    DType val = (cls_target == cls_idx);
    for (int mask_pixel = threadIdx.x; mask_pixel < mask_vol; mask_pixel += blockDim.x) {
      mask_cls_out[mask_pixel] = val;
      ...
    }
```

Comment (Kh4L) [reply src/operator/contrib/mrcnn_target.cu]: Oh my bad, sure I will update it

Comment (Kh4L) [reply src/operator/contrib/mrcnn_target.cu]: Done

File: src/operator/contrib/mrcnn_target.cu (hunk 2 above)
Comment (Kh4L) [reply src/operator/contrib/mrcnn_target.cu]: Done

File: src/operator/contrib/mrcnn_target.cu (hunk 1 above)
Comment (Kh4L) [reply src/operator/contrib/mrcnn_target.cu]: Done

**Pull request merged**